File: core/router.go
package core

import (
	"fmt"
	"math/rand"
	"time"
)

type PeerLink interface {
	PeerID() NodeID
	SendFrame(frame []byte) error
	RSSI() int
	TxPHY() PHY
	RxPHY() PHY
}

type ActionType string

const (
	ActionDeliverLocal ActionType = "deliver-local"
	ActionRelayFlood   ActionType = "relay-flood"
	ActionRelayNextHop ActionType = "relay-next-hop"
	ActionSendAck      ActionType = "send-ack"
	ActionDrop         ActionType = "drop"
)

type Action struct {
	Type        ActionType
	Reason      string
	Datagram    Datagram
	NextHop     *NodeID
	ExcludePeer *NodeID
}

type Router struct {
	LocalID     NodeID
	Identity    *Identity
	Name        string
	Platform    string
	Description string
	dedup       *DedupCache
	Neighbors   *NeighborTable
	Topology    *Topology
	Allowlist   map[NodeID]bool
}

func NewRouter(id NodeID) *Router {
	return &Router{
		LocalID:   id,
		dedup:     NewDedupCache(),
		Neighbors: NewNeighborTable(),
		Topology:  NewTopology(),
		Allowlist: make(map[NodeID]bool),
	}
}

func NewRouterForIdentity(id *Identity) *Router {
	r := NewRouter(id.NodeID())
	r.Identity = id
	return r
}

func (r *Router) HandleDatagram(dg Datagram, incomingPeer *NodeID) []Action {
	if dg.Version != DatagramVersion {
		return r.drop(dg, DropInvalidVersion)
	}
	if dg.TTL < 1 || dg.TTL > MaxTTL || len(dg.Path) > MaxRouteHops || len(dg.ID) != DatagramIDBytes {
		return r.drop(dg, DropMalformed)
	}
	if len(dg.Route) > MaxRouteHops {
		return r.drop(dg, DropBadRoute)
	}
	if incomingPeer != nil && len(r.Allowlist) > 0 && !r.Allowlist[*incomingPeer] {
		return r.drop(dg, DropPeerNotAllowed)
	}
	if r.dedup.SeenOrAdd(dg.ID) {
		return r.drop(dg, DropDuplicate)
	}
	for _, hop := range dg.Path {
		if hop == r.LocalID {
			return r.drop(dg, DropLoop)
		}
	}
	if ok := r.verifyControlIfAnnounce(dg); ok != nil && !*ok {
		return r.drop(dg, DropBadSignature)
	}
	if dg.IsSourceRouted() {
		return r.handleSourceRoute(dg)
	}
	return r.handleFlood(dg, incomingPeer)
}

func (r *Router) verifyControlIfAnnounce(dg Datagram) *bool {
	if dg.Protocol != ProtocolBLEEdgeControl {
		return nil
	}
	ctrl, err := DecodeControl(dg.Payload)
	if err != nil || ctrl.Kind != ControlAnnounce {
		return nil
	}
	body, err := DecodeAnnounceBody(ctrl.Body)
	if err != nil {
		v := false
		return &v
	}
	if len(body.PublicKey) != PublicKeyBytes || NodeIDFromPubKey(body.PublicKey) != dg.Source || !body.Valid() {
		v := false
		return &v
	}
	r.Topology.Update(TopoNode{
		ID:          dg.Source,
		Caps:        body.Caps,
		Neighbors:   body.Neighbors,
		Epoch:       body.Epoch,
		Seq:         body.Seq,
		Timestamp:   body.Timestamp,
		Description: body.Description,
		Name:        body.Name,
		Platform:    body.Platform,
		PublicKey:   body.PublicKey,
	})
	v := true
	return &v
}

func (r *Router) handleFlood(dg Datagram, incomingPeer *NodeID) []Action {
	dg.Path = append(append([]NodeID(nil), dg.Path...), r.LocalID)
	var actions []Action

	if dg.IsBroadcast() || dg.Destination == r.LocalID {
		actions = append(actions, Action{Type: ActionDeliverLocal, Datagram: dg})
		if !dg.IsBroadcast() && dg.AckRequested() {
			actions = append(actions, r.buildAck(dg))
		}
	}

	if dg.TTL > 1 && (dg.IsBroadcast() || dg.Destination != r.LocalID) {
		relay := dg
		relay.TTL--
		actions = append(actions, Action{Type: ActionRelayFlood, Datagram: relay, ExcludePeer: incomingPeer})
	}
	return actions
}

func (r *Router) handleSourceRoute(dg Datagram) []Action {
	cursor := int(dg.RouteCursor)
	if cursor >= len(dg.Route) || len(dg.Route) == 0 {
		return r.drop(dg, DropBadRoute)
	}
	if dg.Route[cursor] != r.LocalID {
		return r.drop(dg, DropNotNextHop)
	}
	if dg.Route[len(dg.Route)-1] != dg.Destination {
		return r.drop(dg, DropBadRoute)
	}
	if int(dg.TTL) != len(dg.Route)-cursor {
		return r.drop(dg, DropBadTTL)
	}
	dg.Path = append(append([]NodeID(nil), dg.Path...), r.LocalID)
	dg.TTL--
	dg.RouteCursor++
	if int(dg.RouteCursor) >= len(dg.Route) {
		actions := []Action{{Type: ActionDeliverLocal, Datagram: dg}}
		if dg.AckRequested() {
			actions = append(actions, r.buildAck(dg))
		}
		return actions
	}
	next := dg.Route[dg.RouteCursor]
	return []Action{{Type: ActionRelayNextHop, Datagram: dg, NextHop: &next}}
}

func (r *Router) buildAck(delivered Datagram) Action {
	route := append([]NodeID(nil), delivered.Path[:len(delivered.Path)-1]...)
	for i, j := 0, len(route)-1; i < j; i, j = i+1, j-1 {
		route[i], route[j] = route[j], route[i]
	}
	route = append(route, delivered.Source)
	payload, _ := AckBody{AckedID: delivered.ID}.ToControl()
	ack := Datagram{
		Version:     DatagramVersion,
		ID:          NewDatagramID(),
		Source:      r.LocalID,
		Destination: delivered.Source,
		TTL:         uint8(len(route)),
		Route:       route,
		Protocol:    ProtocolBLEEdgeControl,
		Payload:     payload,
	}
	r.MarkOriginated(ack.ID)
	next := route[0]
	return Action{Type: ActionSendAck, Datagram: ack, NextHop: &next}
}

// BuildBridged constructs an ACK_BRIDGED control datagram addressed to dst (the original sender of
// a bridged channel message). Delivered by source route when one is known, else flooded.
func (r *Router) BuildBridged(dst NodeID, bridgedID DatagramID, meshHash []byte) (Datagram, bool) {
	payload, err := BridgedBody{BridgedID: bridgedID, BridgeID: r.LocalID, MeshHash: meshHash}.ToControl()
	if err != nil {
		return Datagram{}, false
	}
	return r.NewUnicast(dst, ProtocolBLEEdgeControl, payload, 0, DefaultFloodTTL, false)
}

func (r *Router) BuildAnnounce(caps Capabilities, epoch uint64, seq uint32) (Datagram, error) {
	if r.Identity == nil {
		return Datagram{}, fmt.Errorf("router has no signing identity")
	}
	body := NewAnnounceBody(r.Identity, epoch, seq, time.Now().Unix(), caps, r.Neighbors.IDs(), r.Name, r.Description, r.Platform)
	payload, err := body.ToControl()
	if err != nil {
		return Datagram{}, err
	}
	dg := Datagram{
		Version:     DatagramVersion,
		ID:          NewDatagramID(),
		Source:      r.LocalID,
		Destination: BroadcastNodeID,
		TTL:         AnnounceTTL,
		Protocol:    ProtocolBLEEdgeControl,
		Payload:     payload,
	}
	r.MarkOriginated(dg.ID)
	return dg, nil
}

func (r *Router) NewBroadcast(protocol PayloadProtocol, payload []byte, ttl uint8) Datagram {
	if ttl == 0 || ttl > MaxTTL {
		ttl = DefaultFloodTTL
	}
	dg := Datagram{Version: DatagramVersion, ID: NewDatagramID(), Source: r.LocalID, Destination: BroadcastNodeID, TTL: ttl, Protocol: protocol, Payload: payload}
	r.MarkOriginated(dg.ID)
	return dg
}

func (r *Router) NewUnicast(dst NodeID, protocol PayloadProtocol, payload []byte, flags uint16, floodTTL uint8, requireRoute bool) (Datagram, bool) {
	route := r.SelectRoute(dst)
	var dg Datagram
	if len(route) > 0 {
		dg = Datagram{Version: DatagramVersion, ID: NewDatagramID(), Source: r.LocalID, Destination: dst, TTL: uint8(len(route)), Route: route, Protocol: protocol, Flags: flags, Payload: payload}
	} else {
		if requireRoute {
			return Datagram{}, false
		}
		if floodTTL == 0 || floodTTL > MaxTTL {
			floodTTL = DefaultFloodTTL
		}
		dg = Datagram{Version: DatagramVersion, ID: NewDatagramID(), Source: r.LocalID, Destination: dst, TTL: floodTTL, Protocol: protocol, Flags: flags, Payload: payload}
	}
	r.MarkOriginated(dg.ID)
	return dg, true
}

func (r *Router) SelectRoute(dst NodeID) []NodeID {
	if _, ok := r.Neighbors.Get(dst); ok {
		return []NodeID{dst}
	}
	// Seed the search with our direct neighbors: we're never in our own topology, so a plain BFS
	// from LocalID would dead-end and never find a multi-hop source route.
	all := r.Neighbors.All()
	localNeighbors := make([]NodeID, 0, len(all))
	for _, nb := range all {
		localNeighbors = append(localNeighbors, nb.ID)
	}
	return r.Topology.BFSPathFromSource(r.LocalID, dst, localNeighbors)
}

func (r *Router) MarkOriginated(id DatagramID) {
	r.dedup.SeenOrAdd(id)
}

func (r *Router) FloodJitter() time.Duration {
	return time.Duration(10+rand.Intn(91)) * time.Millisecond
}

func (r *Router) drop(dg Datagram, reason DropReason) []Action {
	return []Action{{Type: ActionDrop, Reason: string(reason), Datagram: dg}}
}

func (r *Router) PublicKeyFor(id NodeID) []byte {
	if n, ok := r.Neighbors.Get(id); ok && len(n.PublicKey) == PublicKeyBytes {
		return append([]byte(nil), n.PublicKey...)
	}
	if n, ok := r.Topology.GetNode(id); ok && len(n.PublicKey) == PublicKeyBytes {
		return append([]byte(nil), n.PublicKey...)
	}
	return nil
}

func (r *Router) NameFor(id NodeID) string {
	if n, ok := r.Topology.GetNode(id); ok {
		if n.Name != "" {
			return n.Name
		}
		return DefaultNodeName(n.PublicKey)
	}
	if n, ok := r.Neighbors.Get(id); ok && len(n.PublicKey) == PublicKeyBytes {
		return DefaultNodeName(n.PublicKey)
	}
	return ""
}

func (r *Router) DescriptionFor(id NodeID) string {
	if n, ok := r.Topology.GetNode(id); ok {
		return n.Description
	}
	return ""
}

func (r *Router) PlatformFor(id NodeID) string {
	if n, ok := r.Topology.GetNode(id); ok {
		return n.Platform
	}
	return ""
}

File: android/sidepath-protocol/src/main/java/cz/meshcore/sidepath/protocol/Router.kt
package cz.meshcore.sidepath.protocol

import kotlin.random.Random

enum class ActionType { DELIVER_LOCAL, RELAY_FLOOD, RELAY_NEXT_HOP, SEND_ACK, DROP }

/**
 * One instruction the transport must execute for an incoming datagram. The
 * [datagram] is the post-processing form (path appended, ttl/cursor updated).
 */
data class Action(
    val type: ActionType,
    val datagram: Datagram,
    val nextHop: NodeId? = null,     // RELAY_NEXT_HOP / SEND_ACK: the single hop to send to
    val excludePeer: NodeId? = null, // RELAY_FLOOD: incoming link to exclude (split horizon)
    val reason: String = "",         // DROP reason
)

object DropReason {
    const val INVALID_VERSION = "invalid-version"
    const val MALFORMED = "malformed"
    const val BAD_TTL = "bad-ttl"
    const val BAD_ROUTE = "bad-route"
    const val PEER_NOT_ALLOWED = "peer-not-allowed"
    const val DUPLICATE = "duplicate"
    const val LOOP = "loop"
    const val NOT_NEXT_HOP = "not-next-hop"
    const val BAD_SIGNATURE = "bad-signature"
}

/**
 * The Sidepath routing engine (docs/PROTOCOL.md §10). Pure logic with no BLE or
 * platform dependencies: it consumes a decoded [Datagram] plus the incoming peer
 * link and returns the [Action]s the transport must execute. Application payloads
 * are opaque; only SIDEPATH_CONTROL announces are inspected (for verification).
 */
class Router(val identity: Identity) {
    val localId: NodeId = identity.nodeId
    val neighbors = NeighborTable()
    val topology = Topology()
    private val dedup = DedupCache()

    /** Optional direct-peer allowlist of hex NodeIDs; empty = allow all. */
    val allowlist: MutableSet<String> = mutableSetOf()

    /** Records [id] as already seen so a flood echo of our own datagram is dropped. */
    fun markOriginated(id: ByteArray) { dedup.seenOrAdd(id) }

    /** Processes an incoming datagram and returns the actions to execute. */
    fun handle(dg: Datagram, incomingPeer: NodeId?): List<Action> {
        // §10.1 common checks
        if (dg.version != Sidepath.DATAGRAM_VERSION) return drop(dg, DropReason.INVALID_VERSION)
        if (dg.id.size != Sidepath.DATAGRAM_ID_BYTES) return drop(dg, DropReason.MALFORMED)
        if (dg.ttl < 1 || dg.ttl > Sidepath.MAX_TTL) return drop(dg, DropReason.BAD_TTL)
        if (dg.path.size > Sidepath.MAX_ROUTE_HOPS) return drop(dg, DropReason.BAD_ROUTE)
        if (dg.route.isNotEmpty() && dg.route.size > Sidepath.MAX_ROUTE_HOPS) return drop(dg, DropReason.BAD_ROUTE)
        if (incomingPeer != null && allowlist.isNotEmpty() && incomingPeer.toHex() !in allowlist) {
            return drop(dg, DropReason.PEER_NOT_ALLOWED)
        }
        if (dedup.seenOrAdd(dg.id)) return drop(dg, DropReason.DUPLICATE)
        if (dg.path.any { it == localId }) return drop(dg, DropReason.LOOP)

        // §10.1.10 verify announces before using or relaying them.
        if (dg.protocol == PayloadProtocol.SIDEPATH_CONTROL) {
            val verified = verifyControlIfAnnounce(dg)
            if (verified == false) return drop(dg, DropReason.BAD_SIGNATURE)
        }

        return if (dg.isSourceRouted) handleSourceRoute(dg, incomingPeer) else handleFlood(dg, incomingPeer)
    }

    /** Returns null if not an announce, true if verified+stored, false if invalid. */
    private fun verifyControlIfAnnounce(dg: Datagram): Boolean? {
        val ctrl = runCatching { ControlMessage.decode(dg.payload) }.getOrNull() ?: return null
        if (ctrl.kind != ControlKind.ANNOUNCE) return null
        val body = runCatching { AnnounceBody.decode(ctrl.body) }.getOrNull() ?: return false
        if (body.publicKey.size != Sidepath.PUBLIC_KEY_BYTES) return false
        val derived = NodeId.fromPublicKey(body.publicKey)
        if (derived != dg.source) return false
        if (!body.isValid()) return false
        topology.update(
            TopoNode(
                id = derived, publicKey = body.publicKey, caps = body.caps, neighbors = body.neighborIds(),
                epoch = body.epoch, seq = body.seq, timestamp = body.timestamp,
                name = body.name, description = body.description, platform = body.platform,
                bridges = body.bridges,
            ),
        )
        return true
    }

    /**
     * Builds a signed ANNOUNCE body for this node, mirroring Go's `Router.BuildAnnounce`. Emits v3 with
     * a `neighbor_info` section (per-link RSSI/PHY/direction/age) once we have any live neighbors;
     * otherwise falls back to the bare-ID v1/v2 layout so a node with no links still emits the smallest
     * announce. [bridges], when present, force at least v2 (and ride along in v3).
     */
    fun buildAnnounceBody(
        caps: Capabilities,
        epoch: Long,
        seq: Long,
        name: String,
        description: String,
        platform: String,
        bridges: List<BridgeAd> = emptyList(),
        timestamp: Long = System.currentTimeMillis() / 1000,
    ): AnnounceBody {
        val infos = neighbors.announceInfo()
        return if (infos.isNotEmpty()) {
            AnnounceBody.createV3(identity, epoch, seq, timestamp, caps, infos, name, description, platform, bridges)
        } else {
            AnnounceBody.create(identity, epoch, seq, timestamp, caps, emptyList(), name, description, platform, bridges)
        }
    }

    private fun handleFlood(dg: Datagram, incomingPeer: NodeId?): List<Action> {
        val actions = mutableListOf<Action>()

        // Local delivery keeps `path` exactly as received: a node never records itself as a hop when
        // it is the recipient, so `path` holds only the intermediate relays (empty = direct).
        val deliverLocally = dg.isBroadcast || dg.destination == localId
        if (deliverLocally) {
            actions += Action(ActionType.DELIVER_LOCAL, dg)
            if (!dg.isBroadcast && dg.ackRequested()) actions += buildAck(dg)
        }

        // §10.2.5 relay if datagram is broadcast or not addressed to local, and TTL allows.
        val relayNeeded = dg.isBroadcast || dg.destination != localId
        if (relayNeeded && dg.ttl > 1) {
            // Relaying onward: append ourselves as an intermediate hop.
            actions += Action(
                ActionType.RELAY_FLOOD,
                dg.copy(path = dg.path + localId, ttl = dg.ttl - 1),
                excludePeer = incomingPeer,
            )
        }
        return actions
    }

    private fun handleSourceRoute(dg: Datagram, @Suppress("UNUSED_PARAMETER") incomingPeer: NodeId?): List<Action> {
        val cursor = dg.routeCursor
        if (cursor >= dg.route.size) return drop(dg, DropReason.BAD_ROUTE)
        if (dg.route[cursor] != localId) return drop(dg, DropReason.NOT_NEXT_HOP)
        if (dg.route.last() != dg.destination) return drop(dg, DropReason.BAD_ROUTE)
        if (dg.ttl != dg.route.size - cursor) return drop(dg, DropReason.BAD_TTL)

        val updated = dg.copy(ttl = dg.ttl - 1, routeCursor = cursor + 1)
        if (updated.routeCursor >= updated.route.size) {
            // We are the destination: deliver with `path` as received (we do not append ourselves).
            val actions = mutableListOf(Action(ActionType.DELIVER_LOCAL, updated))
            if (updated.ackRequested()) actions += buildAck(updated)
            return actions
        }
        // Relaying onward: record ourselves as an intermediate hop.
        val relayed = updated.copy(path = updated.path + localId)
        val nextHop = relayed.route[relayed.routeCursor]
        return listOf(Action(ActionType.RELAY_NEXT_HOP, relayed, nextHop = nextHop))
    }

    /**
     * Builds a source-routed Sidepath ACK for a locally delivered unicast datagram
     * (§9.2). Route = reverse(delivered.path) + [original.source]; `path` is relays-only.
     */
    private fun buildAck(delivered: Datagram): Action {
        // `path` holds only the intermediate relays (we never appended ourselves on delivery), so the
        // return route is the reversed relay list followed by the original source.
        val route = delivered.path.reversed() + delivered.source
        val ack = Datagram(
            id = Datagram.newDatagramId(),
            source = localId,
            destination = delivered.source,
            ttl = route.size,
            route = route,
            routeCursor = 0,
            protocol = PayloadProtocol.SIDEPATH_CONTROL,
            flags = 0,
            payload = AckBody(delivered.id).toControl().encode(),
        )
        dedup.seenOrAdd(ack.id) // we originated it
        return Action(ActionType.SEND_ACK, ack, nextHop = route.first())
    }

    // ---- origination helpers -------------------------------------------------

    /**
     * Builds a broadcast (flood) datagram originated by this node. Marks the id
     * as seen so a flood echo is dropped.
     */
    fun newBroadcast(protocol: Int, payload: ByteArray, ttl: Int = Sidepath.DEFAULT_FLOOD_TTL): Datagram {
        val dg = Datagram(
            source = localId, destination = NodeId.BROADCAST, ttl = ttl.coerceIn(1, Sidepath.MAX_TTL),
            protocol = protocol, payload = payload,
        )
        markOriginated(dg.id)
        return dg
    }

    /**
     * Builds a unicast datagram to [dst]. Uses a known source route when one is
     * available; otherwise falls back to flood unless [requireRoute] is set, in
     * which case it returns null. Marks the id as seen.
     */
    fun newUnicast(
        dst: NodeId,
        protocol: Int,
        payload: ByteArray,
        flags: Int = 0,
        floodTtl: Int = Sidepath.DEFAULT_FLOOD_TTL,
        requireRoute: Boolean = false,
    ): Datagram? {
        val route = selectRoute(dst)
        val dg = if (route != null) {
            Datagram(
                source = localId, destination = dst, ttl = route.size, route = route, routeCursor = 0,
                protocol = protocol, flags = flags, payload = payload,
            )
        } else {
            if (requireRoute) return null
            Datagram(
                source = localId, destination = dst, ttl = floodTtl.coerceIn(1, Sidepath.MAX_TTL),
                protocol = protocol, flags = flags, payload = payload,
            )
        }
        markOriginated(dg.id)
        return dg
    }

    /**
     * Selects a source route to [dst] (§10.4): `[dst]` if it is a direct neighbor,
     * else a BFS path over learned topology, else null (caller may flood).
     */
    fun selectRoute(dst: NodeId): List<NodeId>? {
        if (neighbors.get(dst) != null) return listOf(dst)
        // Seed the search with our direct neighbors: we're never in our own topology, so a plain
        // BFS from localId would dead-end and never find a multi-hop source route.
        val path = topology.bfsPathFromSource(localId, dst, neighbors.all().map { it.id })
        return if (path != null && path.isNotEmpty()) path else null
    }

    fun publicKeyFor(id: NodeId): ByteArray? =
        neighbors.get(id)?.publicKey?.takeIf { it.size == Sidepath.PUBLIC_KEY_BYTES }
            ?: topology.publicKeyFor(id)

    /** Resolves a display name: verified announce name, else deterministic fallback, else "". */
    fun nameFor(id: NodeId): String {
        topology.getNode(id)?.let { tn ->
            if (tn.name.isNotEmpty()) return tn.name
            defaultNodeName(tn.publicKey).takeIf { it.isNotEmpty() }?.let { return it }
        }
        neighbors.get(id)?.publicKey?.let { pk -> defaultNodeName(pk).takeIf { it.isNotEmpty() }?.let { return it } }
        return ""
    }

    /** Random flood relay jitter in milliseconds (§10.2.7). */
    fun floodJitterMs(): Long = Sidepath.FLOOD_JITTER_MIN_MS +
        Random.nextLong(Sidepath.FLOOD_JITTER_MAX_MS - Sidepath.FLOOD_JITTER_MIN_MS + 1)

    private fun drop(dg: Datagram, reason: String) = listOf(Action(ActionType.DROP, dg, reason = reason))
}
